src/libAtomVM/nifs.c

```diff
@@ -5257,6 +5277,61 @@ static term nif_erlang_lists_subtract(Context *ctx, int argc, term argv[])
     free(cons);
     return result;
 }
+
+#ifdef WITH_ZLIB
+static term nif_zlib_compress_1(Context *ctx, int argc, term argv[])
+{
+    UNUSED(argc)
+    term to_compress = argv[0];
+    if (term_is_list(to_compress)) {
+        term to_compress_binary;
+        term result = list_to_binary(to_compress, &to_compress_binary, ctx);
+        if (result != OK_ATOM) {
+            RAISE_ERROR(result);
+        }
+        to_compress = to_compress_binary;
```

Comment:
I don't really like this approach: we are basically creating a binary (with the same content of the iolist that has been given as argument) on the caller heap that will turn into garbage very soon.
Simple approach: factor out a function that allocates and populates a C buffer.
Cool approach: using iolist folding fun for compressing input piece by piece, without copying data to a temporary buffer.